File: domination_telegram/uteis/send_media.py
from database.models.Character import CharacterWaifu, CharacterHusbando
from database.models._types import MediaType
from aiogram.types import Message

async def send_media(
    caption: str,
        
    character: CharacterWaifu | CharacterHusbando=None,
    message: Message = None,
    bot=None,
    chat_id: int | None = None
    ,reply_markup=None,

):
    """
    Envia mídia usando message (reply) ou bot + chat_id.
    Se chat_id for informado, tenta enviar a mensagem para esse chat.
    """

    if not message and not (bot and chat_id):
        raise ValueError("You must provide either message or bot with chat_id.")

    # Caso não tenha personagem, só envia o texto
    if not character:
        if caption:
            if message:
               return await message.answer(caption)
            else:
              return  await bot.send_message(chat_id=chat_id, text=caption, reply_markup=reply_markup)
        return

    # Funções auxiliares para evitar duplicação
    async def send_photo():
        if message:
          return  await message.answer_photo(photo=character.data, caption=caption, reply_markup=reply_markup)
        else:
           return await bot.send_photo(chat_id=chat_id, photo=character.data, caption=caption, reply_markup=reply_markup)

    async def send_video():
        if message:
            return await message.answer_video(video=character.data, caption=caption, reply_markup=reply_markup)
        else:
            return    await bot.send_video(chat_id=chat_id, video=character.data, caption=caption, reply_markup=reply_markup)

    # Envio conforme tipo de mídia
    if character.media_type in [
        MediaType.IMAGE_URL,
        MediaType.IMAGE_FILEID,
        MediaType.IMAGE_BYTES,
        MediaType.IMAGE_FILE
    ]:
      return  await send_photo()

    elif character.media_type in [
        MediaType.VIDEO_BYTES,
        MediaType.VIDEO_FILE,
        MediaType.VIDEO_URL,
        MediaType.VIDEO_FILEID
    ]:
      return  await send_video()

File: Domination-BOT-node/src/utils/sendmedia.ts
import type { MyContext } from "./customTypes.js";
import { MediaType } from "../../generated/prisma/client.js";
import { InputFile, InlineKeyboard } from "grammy";
import type { Character } from "./types.js";
interface ParamsSendMedia {
  chat_id?: string | number | undefined;
  ctx: MyContext | null | undefined;
  per?: Character | null;
  caption?: string;
  reply_markup?: InlineKeyboard;
}

export async function Sendmedia(ParamsSendMedia: ParamsSendMedia) {
  const { chat_id, ctx, per, caption, reply_markup } = ParamsSendMedia;

  // const resolvedChatId = chat_id ?? ctx?.chat?.id;
  const resolvedChatId = chat_id ?? ctx?.chat?.id;

  if (!resolvedChatId) {
    throw new Error("chat_id não fornecido e ctx.chat.id não disponível");
  }

  const api = ctx?.api;

  const options = {
    ...(caption !== undefined && { caption }),
    parse_mode: "HTML" as const,
    ...(reply_markup && { reply_markup }),
  };

  async function sendPhoto(
    photo: any,
    options: {
      reply_markup?: InlineKeyboard;
      parse_mode: "HTML";
      caption?: string;
    },
  ) {
    if (ctx?.replyWithPhoto) return ctx.replyWithPhoto(photo, options);
    return api!.sendPhoto(resolvedChatId!, photo, options);
  }

  async function sendVideo(video: any) {
    if (ctx?.replyWithVideo) return ctx.replyWithVideo(video, options);
    return api!.sendVideo(resolvedChatId!, video, options);
  }

  async function sendText(text: string) {
    if (ctx?.reply)
      return ctx.reply(text, {
        parse_mode: "HTML",
        ...(reply_markup && { reply_markup }),
      });
    return api!.sendMessage(resolvedChatId!, text, {
      parse_mode: "HTML",
      ...(reply_markup && { reply_markup }),
    });
  }

  if (!per) {
    return sendText(caption ?? "");
  }

  const { mediaType: type, media: media } = per;

  try {
    // IMAGENS
    if (type === MediaType.IMAGE_URL || type === MediaType.IMAGE_FILEID) {
      return await sendPhoto(media, options);
    }

    // if (type === MediaType.IMAGE_BASE64 || type === MediaType.IMAGE_BYTES) {
    //   const buffer = Buffer.from(
    //     media,
    //     type === MediaType.IMAGE_BASE64 ? "base64" : "binary",
    //   );
    //   return await sendPhoto(new InputFile(buffer));
    // }

    // if (type === MediaType.IMAGE_FILE) {
    //   return await sendPhoto(new InputFile(media));
    // }

    // VIDEOS
    if (type === MediaType.VIDEO_URL || type === MediaType.VIDEO_FILEID) {
      return await sendVideo(media);
    }

    // if (type === MediaType.VIDEO_BASE64 || type === MediaType.VIDEO_BYTES) {
    //   const buffer = Buffer.from(
    //     media,
    //     type === MediaType.VIDEO_BASE64 ? "base64" : "binary",
    //   );
    //   return await sendVideo(new InputFile(buffer));
    // }

    // if (type === MediaType.VIDEO_FILE) {
    //   return await sendVideo(new InputFile(media));
    // }

    return sendText("Tipo de mídia não suportado.");
  } catch (error) {
    console.error("Erro ao enviar mídia:", error);
    return sendText(caption ?? "");
  }
}
